MaxText/tests/attention_test.py

```diff
@@ -49,149 +46,172 @@ def setUp(self):
     self.mesh = Mesh(devices_array, self.cfg.mesh_axes)
 
     self.global_batch_size = self.cfg.global_batch_size_to_train_on
-    self.num_heads = self.cfg.base_num_heads
+    self.num_kv_heads = self.cfg.num_kv_heads
+    self.num_query_heads = self.cfg.num_query_heads
     self.max_target_length = self.cfg.max_target_length
     self.head_dim = self.cfg.head_dim
     self.embed_dim = self.cfg.base_emb_dim
     self.dtype = self.cfg.dtype
 
-    self.mha_attention = MultiHeadDotProductAttention(
-        num_heads=self.num_heads,
+    self._attention_as_mha_generic = Attention(
+        num_query_heads=self.num_query_heads,
+        num_kv_heads=self.num_kv_heads,
         head_dim=self.head_dim,
         mesh=self.mesh,
+        attention_kernel = "dot_product",
         dtype=self.dtype,
         dropout_rate=self.cfg.dropout_rate,
         name='self_attention',
     )
-    self.variable = self.mha_attention.init(
+
+    self._attention_as_mha_generic_variable = self._attention_as_mha_generic.init(
         {'params': self.rng, 'aqt': self.rng},
         jnp.ones(
             (self.global_batch_size, self.max_target_length, self.embed_dim)),
         jnp.ones(
             (self.global_batch_size, self.max_target_length, self.embed_dim)),
-        'mha',
+        jnp.ones(
+            (self.global_batch_size, self.max_target_length)),
     )
 
-  def get_decoder_mask(self, dtype):
-    a = jnp.stack([
-        jnp.tri(self.max_target_length, dtype=dtype)[jnp.newaxis, :]
-        for _ in range(self.global_batch_size)
-    ])
-    return a
+    self._attention_as_mha_flash = Attention(
```

Comment:
No need to create this and following in setup as they are not always needed for different tests.

Reply:
Wow, I thought that __init__ was shared but it isn't! Good to know!